File: java_basics/src/test/java/com/user/selenium/listeners/ScreenshotListener.java
package com.user.selenium.listeners;

import com.user.selenium.BaseTest;
import com.user.selenium.utils.LoggerUtil;
import org.openqa.selenium.OutputType;
import org.openqa.selenium.TakesScreenshot;
import org.openqa.selenium.WebDriver;
import org.testng.ITestListener;
import org.testng.ITestResult;

import java.io.File;
import java.nio.file.Files;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.logging.Logger;

public class ScreenshotListener implements ITestListener {

    private static final Logger logger = LoggerUtil.getLogger();
    private static final String SCREENSHOTS_DIR = "results/screenshots";

    @Override
    public void onTestFailure(ITestResult result) {
        Object testClass = result.getInstance();

        if (!(testClass instanceof BaseTest)) {
            return;
        }

        WebDriver driver = ((BaseTest) testClass).getDriver();

        if (driver == null) {
            logger.warning("WebDriver was null, screenshot not captured");
            return;
        }

        try {
            String timestamp = LocalDateTime.now()
                    .format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss"));
            String testName = result.getMethod().getMethodName();
            String filename = testName + "_" + timestamp + ".png";

            File srcFile = ((TakesScreenshot) driver).getScreenshotAs(OutputType.FILE);
            File destFile = new File(SCREENSHOTS_DIR, filename);

            Files.copy(srcFile.toPath(), destFile.toPath());

            logger.warning("Test failed. Screenshot saved to: " + destFile.getPath());

        } catch (Exception e) {
            logger.severe("Failed to capture screenshot: " + e.getMessage());
        }
    }
}

File: kotlin_basics/src/test/kotlin/listeners/ScreenshotListener.kt
package listeners

import base.BaseTest
import org.openqa.selenium.OutputType
import org.openqa.selenium.TakesScreenshot
import org.testng.ITestListener
import org.testng.ITestResult
import java.io.File
import java.nio.file.Files
import java.nio.file.StandardCopyOption

class ScreenshotListener : ITestListener {

    override fun onTestFailure(result: ITestResult) {
        val testInstance = result.instance

        if (testInstance is BaseTest) {
            val screenshot = (testInstance.driver as TakesScreenshot)
                .getScreenshotAs(OutputType.FILE)

            val screenshotFile = File(
                "results/screenshots/${result.name}.png"
            )

            Files.copy(
                screenshot.toPath(),
                screenshotFile.toPath(),
                StandardCopyOption.REPLACE_EXISTING
            )
        }
    }
}
